File: 3-Structural Patterns/1-Flyweight Pattern/BMW Model Example/Python/flyweight_pattern_test.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Application module.
"""

__author__ = 'Ziang Lu'

from flyweight.customization import BMWCarCustomization
from flyweight.flyweight_car_factory import FlyweightBMWCarFactory, Model


def main():
    flyweight_bmw_car_factory = FlyweightBMWCarFactory()

    # In real-world application, we need to create a large number of objects of
    # similar nature, which would have consumed a large, unaccepted amount of
    # memory and decrease the performance.
    #
    # By using Flyweight Pattern, we first get the shared  "ConcreteFlyweight"
    # from the "FlyweightFactory", and then customize it to get the
    # "UnsharedConcreteFlyweight" we desire.
    # In this way, the number of physically created objects is greatly reduced,
    # and thus the load on memory is reduced and the  performance is improved.

    # Get the shared BMWSeries1 from the FlyweightBMWCarFactory
    series1car = flyweight_bmw_car_factory.get_flyweight_bmw_car(Model.Series1)
    series1car2 = flyweight_bmw_car_factory.get_flyweight_bmw_car(Model.Series1)
    # Make sure series1Car and series1Car2 are the same instance
    print(series1car)
    print(series1car2)

    print()

    # Get the shared BMWSeries2 from the FlyweightBMWCarFactory
    series2car = flyweight_bmw_car_factory.get_flyweight_bmw_car(Model.Series2)
    series2car2 = flyweight_bmw_car_factory.get_flyweight_bmw_car(Model.Series2)
    # Make sure series2Car and series2Car2 are the same instance
    print(series2car)
    print(series2car2)

    customization1 = BMWCarCustomization(tire_size=19,
                                         laser_signature='Oh yeah')
    customization2 = BMWCarCustomization(tire_size=21,
                                         laser_signature='Give you a ride!')

    print()
    print('Series 1 + Customization 1:')
    series1car.print_full_characteristics(customization1)
    print(f'Full price: {series1car.calc_price(customization1)}')

    print()
    print('Series 1 + Customization 2:')
    series1car.print_full_characteristics(customization2)
    print(f'Full price: {series1car.calc_price(customization2)}')

    print()
    print('Series 2 + Customization 1:')
    series2car.print_full_characteristics(customization1)
    print(f'Full price: {series2car.calc_price(customization1)}')

    print()
    print('Series 2 + Customization 2:')
    series2car.print_full_characteristics(customization2)
    print(f'Full price: {series2car.calc_price(customization2)}')


if __name__ == '__main__':
    main()

# Output:
# <flyweight.car.BMWSeries1 object at 0x108b52af0>
# <flyweight.car.BMWSeries1 object at 0x108b52af0>
#
# <flyweight.car.BMWSeries2 object at 0x108b52b00>
# <flyweight.car.BMWSeries2 object at 0x108b52b00>
#
# Series 1 + Customization 1:
# Tire Size: 19
# Laser Signature: Oh yeah
# Laser Signature: 7
# Full price: 27187
#
# Series 1 + Customization 2:
# Tire Size: 21
# Laser Signature: Give you a ride!
# Laser Signature: 16
# Full price: 28087
#
# Series 2 + Customization 1:
# Tire Size: 19
# Laser Signature: Oh yeah
# Laser Signature: 7
# Full price: 30400
#
# Series 2 + Customization 2:
# Tire Size: 21
# Laser Signature: Give you a ride!
# Laser Signature: 16
# Full price: 32200

File: 3-Structural Patterns/1-Flyweight Pattern/BMW Model Example/Java/runner/FlyweightPatternTest.java
package runner;

import flyweight.FlyweightBMWCarFactory;
import flyweight.car.FlyweightBMWCar;
import flyweight.customization.BMWCarCustomization;

/**
 * Application that actually uses Flyweight Pattern.
 *
 * @author Ziang Lu
 */
public class FlyweightPatternTest {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        FlyweightBMWCarFactory flyweightBMWCarFactory = FlyweightBMWCarFactory.getInstance();

        /*
         * In real-world production, we need to create a large number of objects of similar nature, which would have
         * consumed a large, unaccepted amount of memory and decrease the performance.
         *
         * By using Flyweight Pattern, we first get the shared "ConcreteFlyweight" from the "FlyweightFactory", and then
         * customize it to get the "UnsharedConcreteFlyweight" we desire.
         * In this way, the number of physically created objects is greatly reduced, and thus the load on memory is
         * reduced and the performance is improved.
         */

        // Get the shared BMWSeries1 from the FlyweightBMWCarFactory
        FlyweightBMWCar series1Car = flyweightBMWCarFactory.getFlyweightBMWCar(FlyweightBMWCarFactory.Model.Series1);
        FlyweightBMWCar series1Car2 = flyweightBMWCarFactory.getFlyweightBMWCar(FlyweightBMWCarFactory.Model.Series1);
        // Make sure series1Car and series1Car2 are the same instance
        System.out.println(series1Car);
        System.out.println(series1Car2);

        System.out.println();

        // Get the shared BMWSeries2 from the FlyweightBMWCarFactory
        FlyweightBMWCar series2Car = flyweightBMWCarFactory.getFlyweightBMWCar(FlyweightBMWCarFactory.Model.Series2);
        FlyweightBMWCar series2Car2 = flyweightBMWCarFactory.getFlyweightBMWCar(FlyweightBMWCarFactory.Model.Series2);
        // Make sure series2Car and series2Car2 are the same instance
        System.out.println(series2Car);
        System.out.println(series2Car2);

        BMWCarCustomization customization1 = new BMWCarCustomization(19, "Oh yeah");
        BMWCarCustomization customization2 = new BMWCarCustomization(21, "Give you a ride!");

        System.out.println();
        System.out.println("Series 1 + Customization 1:");
        series1Car.printFullCharacteristics(customization1);
        System.out.println("Full price: " + series1Car.calcPrice(customization1));

        System.out.println();
        System.out.println("Series 1 + Customization 2:");
        series1Car.printFullCharacteristics(customization2);
        System.out.println("Full price: " + series1Car.calcPrice(customization2));

        System.out.println();
        System.out.println("Series 2 + Customization 1:");
        series2Car.printFullCharacteristics(customization1);
        System.out.println("Full price: " + series2Car.calcPrice(customization1));

        System.out.println();
        System.out.println("Series 2 + Customization 2:");
        series2Car.printFullCharacteristics(customization2);
        System.out.println("Full price: " + series2Car.calcPrice(customization2));

        /*
         * Output:
         * flyweight.car.BMWSeries1@387c703b
         * flyweight.car.BMWSeries1@387c703b
         *
         * flyweight.car.BMWSeries1@387c703b
         * flyweight.car.BMWSeries1@387c703b
         *
         * Series 1 + Customization 1:
         * Tire Size: 19
         * Laser Signature: Oh yeah
         * Laser Signature length: 7
         * Full price: 27187
         *
         * Series 1 + Customization 2:
         * Tire Size: 21
         * Laser Signature: Give you a ride!
         * Laser Signature length: 16
         * Full price: 28087
         *
         * Series 2 + Customization 1:
         * Tire Size: 19
         * Laser Signature: Oh yeah
         * Laser Signature length: 7
         * Full price: 30400
         *
         * Series 2 + Customization 2:
         * Tire Size: 21
         * Laser Signature: Give you a ride!
         * Laser Signature length: 16
         * Full price: 32200
         */
    }

}
